devito/finite_differences/coefficients.py

```diff
@@ -0,0 +1,191 @@
+                                           side=side, stagger=stagger)
+
+            for j in range(len(coeffs)):
+                subs.update({function.fd_coeff_symbol()
+                             (indices[j], deriv_order, function, dim): coeffs[j]})
+
+            return subs
+
+        # Figure out when symbolic coefficients can be replaced
+        # with user provided coefficients and, if possible, generate
+        # replacement rules
+        rules = {}
+        for d in self.data:
+            if isinstance(d.coefficients, np.ndarray):
+                rules.update(generate_subs(d))
+
+        return rules
+
+
+class Coefficient(object):
+    """
+    Prepare custom FD coefficients to pass to Coefficients object.
+
+    Parameters
+    ----------
+    deriv_order : represents the order of the derivative being taken.
+    function : represents the function for which the supplied coefficients
+               will be used.
+    dimension : represents the dimension with respect to which the
+                derivative is being taken.
+    coefficients : represents the set of finite difference coefficients
+                  intended to be used in place of the standard
+                  coefficients (obtained from a Taylor expansion).
+    """
+
+    def __init__(self, deriv_order, function, dimension, coefficients):
+
+        self.check_input(deriv_order, function, dimension, coefficients)
+
+        # Ensure the given set of coefficients is the correct length
+        if dimension.is_Time:
+            if len(coefficients)-1 != function.time_order:
+                raise ValueError("Number FD weights provided does not "
+                                 "match the functions space_order")
+        elif dimension.is_Space:
+            if len(coefficients)-1 != function.space_order:
+                raise ValueError("Number FD weights provided does not "
+                                 "match the functions space_order")
+
+        self.is_Coefficient = True
+
+        self.deriv_order = deriv_order
+        self.function = function
+        self.dimension = dimension
+        self.coefficients = coefficients
+
+    def check_input(self, deriv_order, function, dimension, coefficients):
+        if not isinstance(deriv_order, int):
+            raise TypeError("Derivative order must be an integer")
+        if not function.is_Function:
+            raise TypeError("Coefficients must be attached to a valid function")
+        if not dimension.is_Dimension:
+            raise TypeError("Coefficients must be attached to a valid dimension")
+        # Currently only numpy arrays are accepted here.
+        # Functionality will be expanded in the near future.
+        if not isinstance(coefficients, np.ndarray):
+            raise NotImplementedError
+        return
+
+
+def default_rules(obj, functions):
+
+    def generate_subs(d):
+
+        deriv_order = d[0]
```

Comment:
I prefer an explicit parameter list

Reply:
I'm not sure it's needed here since this is a local function? The unpacking would just be moved to the for loop which would make things a bit untidier and not really add any clarity.

Reply:
the point is, this kind of unpacking doesn't help with code self-documentation. Here, ``d`` could contain 100 parameters for example.

Another example: in the ``rules`` method above, ``d`` actually contains 4 parameters? as in, it includes ``coeffs`` as well

I'm not sure I got this:
> The unpacking would just be moved to the for loop
 
do you mean the caller? what line?